### DISCUSSION TITLE
improvements regarding the new sec edgar API

### DISCUSSION THREAD

**Discussion by tdktdkett** [sec-edgar/sec-edgar#186] [Ideas]
Hey guys. 
I haven't been working on this for a while and it looks like there are some new updates from their API which we can definitely levarage to accelerate. 
So it seems there's a new filling search endpoint:
```
/submissions/CIK{cik}.json
```
for example 
```
https://data.sec.gov/submissions/CIK0001006240.json
```
and it has the following key in the dict
- accessionNumber
- primaryDocument
- form
json['fillings']['recent'].keys() you'll understand
```
https://www.sec.gov/Archives/edgar/data/cik/accessionNumber/primaryDocument
https://www.sec.gov/Archives/edgar/data/0001006240/000119312513244630/d548819dsctoia.htm
```
I'm proposing a alternative approach that for specific company we may just 
1. Pull this json, parse it
2. Build cache/db with it. 

Let me know if anyone has already been working on it, otherwise I'll just shot a PR this weekend.

**Comment by jackmoody11** [comment]
That would be great! Thanks @tdktdkett.

**Comment by jackmoody11** [comment]
@tdktdkett how did you find out about this new way of getting filings for a CIK?

**Reply by tdktdkett** [reply]
I was poking around the redirected page https://www.sec.gov/cgi-bin/browse-edgar?CIK=0001006240&action=getcompany. 
It seems to have a nicely formated json response

**Reply by reteps** [reply]
This API is limited to 10,000 max responses.

**Reply by wayneroe** [reply]
@reteps Have you tried it with a user-agent?
It seems to work when supplying a user-agent header. However, I haven't tried it 10'000 times yet.
```python
import time
import requests
for i in range(100):
    time.sleep(0.1)
    r = requests.get("https://data.sec.gov/submissions/CIK0001006240.json", headers={"User-agent": "wayne.roettger@gmail.com", 'From': 'wayne.roettger@gmail.com'})
    if r.status_code == 403:
        print("Request Rate Threshold Exceeded!")
```

**Comment by dtrianab** [comment]
> 
> 
> Hey guys.
> I haven't been working on this for a while and it looks like there are some new updates from their API which we can definitely levarage to accelerate.
> So it seems there's a new filling search endpoint:
> 
> ```
> /submissions/CIK{cik}.json
> ```
> 
> for example
> 
> ```
> https://data.sec.gov/submissions/CIK0001006240.json
> ```
> 
> and it has the following key in the dict
> 
>     * accessionNumber
> 
>     * primaryDocument
> 
>     * form
>       json['fillings']['recent'].keys() you'll understand
> 
> 
> ```
> https://www.sec.gov/Archives/edgar/data/cik/accessionNumber/primaryDocument
> https://www.sec.gov/Archives/edgar/data/0001006240/000119312513244630/d548819dsctoia.htm
> ```
> 
> I'm proposing a alternative approach that for specific company we may just
> 
>     1. Pull this json, parse it
> 
>     2. Build cache/db with it.
> 
> 
> Let me know if anyone has already been working on it, otherwise I'll just shot a PR this weekend.

Hi there, 
Im trying to use the API as well, seems like at the moment is for beta testing, so some errors could appear.  https://www.sec.gov/edgar/sec-api-documentation
I have run below code, I noticed from that example in ALCATEL that there is no dash in the sec url, however there is some dash in accessionNumber received in json response, that had to be removed. I placed the json response into a dataframe, if I look at last row, I can go to the record ( https://www.sec.gov/Archives/edgar/data/0000884887/000114036112008835/xslF345X03/doc1.xml ) filed in year 2012, which covers quite long span, but I suspect there are more before this, any idea if those before could be queried ?

```python
import json
import requests
import pandas as pd

#CIK number of a company (RCL example)
cik_num = '0000884887'

#Submissions Endpoint
sub = 'https://data.sec.gov/submissions/'
url_sub = sub+'CIK'+cik_num+'.json'
response_sub = requests.get(url_sub, headers={"User-agent": "dtb_fk@hotmail.com", 'From': 'dtb_fk@hotmail.com'})
json_data_sub = json.loads(response_sub.text)
filings=pd.DataFrame.from_dict(json_data_sub['filings']['recent'])
merge_col_data = 'https://www.sec.gov/Archives/edgar/data/'+cik_num+'/'+filings.accessionNumber[0]+'/'+filings.primaryDocument[0]
sec_url = merge_col_data.replace('-','')
print(sec_url)
```
```python
print(sec_url)
https://www.sec.gov/Archives/edgar/data/0000884887/000110465921103667/tm2124790d2_8k.htm
```

**Reply by mpreiss9** [reply]
I wonder if you couldn't spare yourself all the online connecting work by just downloading the bulk file. Per the SEC:
Bulk data
For bulk data users, the data contained in both the Frame and the Company Facts APIs can be downloaded in a single archive file, which is updated nightly:

http://www.sec.gov/Archives/edgar/daily-index/xbrl/companyfacts.zip

I did, and it's only about a gig in size. The zip has one json file for every CIK, named CIK0000000000.json. The contents cover all the years XBRL filings have been made (10 +/-) for every company (almost 15,000 jsons). It does seem to only include the items in the us-gaap taxonomy, but then so does the api if I understand correctly. Nevertheless, for a typical company like CSCO there are about 750 tags x 11 years plus every quarter for those 11 years. All for the price of a 10 second download. In any event, it's a great source for testing data ingestion even if you later want to use the api anyway. The way the data is set up is a little cumbersome, so somebody needs to figure out how to transform it so it's easier to work with. I'm a novice but played around enough to understand the general structure I think

**Reply by jackmoody11** [reply]
This is really interesting @mpreiss9. Would you be interested in having a go at a PR? I am happy to go through the code review process with you.

**Reply by jackmoody11** [reply]
@dtrianab if you are still interested, working with @mpreiss9 could make this a really good addition.

**Reply by ThePythonDude** [reply]
Hello @jackmoody11 , 

I have played the whole summer with the .json file of the bulk .zip file. Unfortunately, it is true that custom tags are not included in these files, what makes them not really useful in my opinion. I have already contacted the support of the Edgar API and provided feedback in that regard. However, I haven't received an answer.

When they include the custom tags, then the next challenge would be to hand-map the different tags to a standard term. However, it is not enough to map the us-gaap_tag to a standardized tag, because in one filing the us-gaap_tag could be a sub-term of a main us-gaap tag and in another the us-gaap_tag by itself is the main tag.
To fix the problem, it is to my current best knowledge necessary to map the us-gaap_tag together with its umbrella us-gaap_tag, which can be taken from the cal.xml files, to a standardized tag. 

For now, I will pause my work until the custom tags are hopefully added to the .json files in the bulk. zip file.

**Reply by jackmoody11** [reply]
@ThePythonDude I haven't taken a look at the bulk file yet. I wonder how hard it would be to map to the specific tags like you mentioned.
